### ISSUE TITLE
Updating required fields for a CRD in the Cluster, then applying changes to a CR later does not work

### ISSUE THREAD

**Issue by anderssonw** [hashicorp/terraform-provider-kubernetes#1923]
### Terraform Version, Provider Version and Kubernetes Version

```
Terraform version: (run-terraform github action v2.03), but same error with local v1.3.4
Kubernetes provider version: v2.16.0
Kubernetes version: v1.24.7-gke.1400 (Anthos On-Prem)

```

### Affected Resource(s)

CustomResourceDefinition, file for CRD is here: https://github.com/kartverket/skiperator/blob/0f64b53f57c126c1ce516f1c68dc243d9be85a81/deployment/skiperator.kartverket.no_applications.yaml

### Terraform Configuration Files
```hcl
resource "kubernetes_manifest" "test-application" {
  manifest = {
    apiVersion = "skiperator.kartverket.no/v1alpha1"
    kind       = "Application"
    metadata = {
      name      = "shipwreck-test-2"
      namespace = "shipwreck"
    }
    spec = {
      image = var.image
      port  = 6379
      replicas = {
        min = 1
        max = 1
      }
      resources = {
        limits = {
          memory           = "1G"
          cpu              = "20m"
          ephemeralStorage = "1G"
          storage          = "1G"
        }
        requests = {
          cpu              = "10m"
          memory           = "500M"
          ephemeralStorage = "500M"
          storage          = "500M"
        }
      }
    }
  }
}

```

### Debug Output

https://github.com/kartverket/shipwreck/actions/runs/3602444385/jobs/6069419784

### Steps to Reproduce

1. Deploy new CRD in cluster
2. Apply changes to manifest to adhere to the new changes (cpu/memory requests are required, memory limit is required)
3. Deploy new Application with new manifest
4. Error

### Expected Behavior
The application was deployed with changes

### Actual Behavior
Provider crashes due to some field constraints of some sort

### Important Factoids

- Applying these changes with the same manifest work just fine when doing this through the kubectl provider.
- The changes being made to the CRD include creating our own definition of `spec.resources`, to be able to apply custom validation. The name for this definition is the same, but the resulting format of the CRD is not quite similar.
For example, the CRD now uses `properties` instead of `additionalProperties` under `limits` and `requests`.
- If the application being made changes to is deleted on the cluster, the provider still crashes due to a discrepancy between the terraform state and the CRD.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by alexsomesan** [comment]
I cannot access the gist you shared. Can you double check that it's accessible?
Alternatively, can you share what the actual error was?

**Comment by anderssonw** [comment]
Sorry, @alexsomesan, I thought it was from our open source project, but it was a different one. Here it is:

```
Error: Plugin did not respond

  with kubernetes_manifest.test-application,
  on main.tf line 5, in resource "kubernetes_manifest" "test-application":
   5: resource "kubernetes_manifest" "test-application" {

The plugin encountered an error, and failed to respond to the
plugin.(*GRPCProvider).UpgradeResourceState call. The plugin logs may contain
more details.

Stack trace from the terraform-provider-kubernetes_v2.16.0_x5 plugin:

panic: can't create a tftypes.Value of type tftypes.Object["cpu":tftypes.String, "ephemeralStorage":tftypes.String, "memory":tftypes.String, "storage":tftypes.String], required attribute "cpu" not set

goroutine 113 [running]:
github.com/hashicorp/terraform-plugin-go/tftypes.NewValue(...)
	github.com/hashicorp/terraform-plugin-go@v0.14.0/tftypes/value.go:273
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.morphMapToType({{0x20ed8f0?, 0xc000b58b40?}, {0x1b5d2e0?, 0xc000b58900?}}, {0x20ed948?, 0xc00552e8d0?}, 0xc001543ef0)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:542 +0x13a5
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.ValueToType({{0x20ed8f0?, 0xc000b58b40?}, {0x1b5d2e0?, 0xc000b58900?}}, {0x20ed948, 0xc00552e8d0}, 0xc001543ef0)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:44 +0x6f9
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.morphObjectToType({{0x20ed948?, 0xc000b59260?}, {0x1b5d2e0?, 0xc000b582d0?}}, {0x20ed948?, 0xc0055789f0?}, 0xc001543ec0)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:605 +0x1079
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.ValueToType({{0x20ed948?, 0xc000b59260?}, {0x1b5d2e0?, 0xc000b582d0?}}, {0x20ed948, 0xc0055789f0}, 0xc001543ec0)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:46 +0x6b5
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.morphObjectToType({{0x20ed948?, 0xc000be23f0?}, {0x1b5d2e0?, 0xc000b1a780?}}, {0x20ed948?, 0xc00588be30?}, 0xc0015436e0)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:605 +0x1079
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.ValueToType({{0x20ed948?, 0xc000be23f0?}, {0x1b5d2e0?, 0xc000b1a780?}}, {0x20ed948, 0xc00588be30}, 0xc0015436e0)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:46 +0x6b5
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.morphObjectToType({{0x20ed948?, 0xc000bf6c60?}, {0x1b5d2e0?, 0xc000b12600?}}, {0x20ed948?, 0xc00596a8d0?}, 0xc001543290)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:605 +0x1079
github.com/hashicorp/terraform-provider-kubernetes/manifest/morph.ValueToType({{0x20ed948?, 0xc000bf6c60?}, {0x1b5d2e0?, 0xc000b12600?}}, {0x20ed948, 0xc00596a8d0}, 0xc001543290)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/morph/morph.go:46 +0x6b5
github.com/hashicorp/terraform-provider-kubernetes/manifest/provider.(*RawProviderServer).UpgradeResourceState(0xc000beaf00, {0x20e7dc0, 0xc000a018f0}, 0xc001323000)
	github.com/hashicorp/terraform-provider-kubernetes/manifest/provider/upgrade_state.go:89 +0x9e7
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.UpgradeResourceState({0xc000e741e0, 0xc000e74240, {0xc000c01bc0, 0x2, 0x2}, 0xc000e74210, 0xc000e46a80, 0xc0010725a0, 0xc000e74270}, {0x20e7dc0, ...}, ...)
	github.com/hashicorp/terraform-plugin-mux@v0.7.0/tf5muxserver/mux_server_UpgradeResourceState.go:27 +0x142
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).UpgradeResourceState(0xc00041de00, {0x20e7dc0?, 0xc000a006f0?}, 0xc000935f40)
	github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/tf5server/server.go:716 +0x270
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_UpgradeResourceState_Handler({0x1d08800?, 0xc00041de00}, {0x20e7dc0, 0xc000a006f0}, 0xc00133cfc0, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:313 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0009f4000, {0x20eded8, 0xc000f78000}, 0xc0011acea0, 0xc001169380, 0x2f10720, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1295 +0xb2b
google.golang.org/grpc.(*Server).handleStream(0xc0009f4000, {0x20eded8, 0xc000f78000}, 0xc0011acea0, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1[63](https://github.com/kartverket/shipwreck/actions/runs/3602444385/jobs/6069419784#step:11:64)6 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.48.0/server.go:[93](https://github.com/kartverket/shipwreck/actions/runs/3602444385/jobs/6069419784#step:11:94)2 +0x[98](https://github.com/kartverket/shipwreck/actions/runs/3602444385/jobs/6069419784#step:11:99)
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.48.0/server.go:930 +0x28a

Error: The terraform-provider-kubernetes_v2.16.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

**Comment by anderssonw** [comment]
Hi @alexsomesan, have you had a chance to look into this?

**Comment by anderssonw** [comment]
@alexsomesan ? Don't mean to be a nuisance just curious seeing as you asked for the gist :)

**Comment by alexsomesan** [comment]
Sorry for the radio silence. That's a strange error message you're getting there. 
I need to try to reproduce it in order to understand what's going on.

**Comment by alexsomesan** [comment]
Unfortunately, I'm not able to reproduce it. I tried your configuration and it worked as expected.

Here's what I got:
```
➜  test-skiperator terraform apply -auto-approve

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # kubernetes_manifest.test-application will be created
  + resource "kubernetes_manifest" "test-application" {
      + manifest = {
          + apiVersion = "skiperator.kartverket.no/v1alpha1"
          + kind       = "Application"
          + metadata   = {
              + name      = "shipwreck-test-2"
              + namespace = "shipwreck"
            }
          + spec       = {
              + image     = "nginx"
              + port      = 80
              + replicas  = {
                  + max = 1
                  + min = 1
                }
              + resources = {
                  + limits   = {
                      + cpu              = "20m"
                      + ephemeralStorage = "1G"
                      + memory           = "1G"
                      + storage          = "1G"
                    }
                  + requests = {
                      + cpu              = "10m"
                      + ephemeralStorage = "500M"
                      + memory           = "500M"
                      + storage          = "500M"
                    }
                }
            }
        }
      + object   = {
          + apiVersion = "skiperator.kartverket.no/v1alpha1"
          + kind       = "Application"
          + metadata   = {
              + annotations                = (known after apply)
              + creationTimestamp          = (known after apply)
              + deletionGracePeriodSeconds = (known after apply)
              + deletionTimestamp          = (known after apply)
              + finalizers                 = (known after apply)
              + generateName               = (known after apply)
              + generation                 = (known after apply)
              + labels                     = (known after apply)
              + managedFields              = (known after apply)
              + name                       = "shipwreck-test-2"
              + namespace                  = "shipwreck"
              + ownerReferences            = (known after apply)
              + resourceVersion            = (known after apply)
              + selfLink                   = (known after apply)
              + uid                        = (known after apply)
            }
          + spec       = {
              + accessPolicy   = {
                  + inbound  = {
                      + rules = (known after apply)
                    }
                  + outbound = {
                      + external = (known after apply)
                      + rules    = (known after apply)
                    }
                }
              + command        = (known after apply)
              + env            = (known after apply)
              + envFrom        = (known after apply)
              + filesFrom      = (known after apply)
              + gcp            = {
                  + auth = {
                      + serviceAccount = (known after apply)
                    }
                }
              + image          = "nginx"
              + ingresses      = (known after apply)
              + labels         = (known after apply)
              + liveness       = {
                  + failureThreshold = (known after apply)
                  + initialDelay     = (known after apply)
                  + path             = (known after apply)
                  + port             = (known after apply)
                  + timeout          = (known after apply)
                }
              + port           = 80
              + readiness      = {
                  + failureThreshold = (known after apply)
                  + initialDelay     = (known after apply)
                  + path             = (known after apply)
                  + port             = (known after apply)
                  + timeout          = (known after apply)
                }
              + replicas       = {
                  + max                  = 1
                  + min                  = 1
                  + targetCpuUtilization = (known after apply)
                }
              + resourceLabels = (known after apply)
              + resources      = {
                  + limits   = {
                      + "cpu"              = "20m"
                      + "ephemeralStorage" = "1G"
                      + "memory"           = "1G"
                      + "storage"          = "1G"
                    }
                  + requests = {
                      + "cpu"              = "10m"
                      + "ephemeralStorage" = "500M"
                      + "memory"           = "500M"
                      + "storage"          = "500M"
                    }
                }
              + startup        = {
                  + failureThreshold = (known after apply)
                  + initialDelay     = (known after apply)
                  + path             = (known after apply)
                  + port             = (known after apply)
                  + timeout          = (known after apply)
                }
              + strategy       = {
                  + type = (known after apply)
                }
            }
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
kubernetes_manifest.test-application: Creating...
kubernetes_manifest.test-application: Creation complete after 0s

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

Here are the versions I used for the test:
```
Terraform v1.3.7
on darwin_arm64
+ provider registry.terraform.io/hashicorp/kubernetes v2.17.0

kubectl version --short
Client Version: v1.26.1
Kustomize Version: v4.5.7
Server Version: v1.25.3

https://github.com/kartverket/skiperator@latest
```

**Comment by anderssonw** [comment]
No worries for the radio silence :)

Creating an application is never an issue. I believe you misunderstood my explanation.

The case was that we already had an application applied in the cluster, with some of the resource fields set. After this, we changed the CRD and the API, where the resource field did now had some required fields.

After applying these changes to the CRD, attempting to modify an already existing application with terraform failed in the manner i supplied in the original issue.


In general we've had some issues with updating the CRD. For example, another issue one of our teams had was that they had defined `spec.strategy`, but then removed it. Removing it caused the following error:

```
│ Error: API response status: Failure
│ 
│   with kubernetes_manifest.nrl-brukeradministrasjon_application,
│   on application.tf line 1, in resource "kubernetes_manifest" "nrl-brukeradministrasjon_application":
│    1: resource "kubernetes_manifest" "nrl-brukeradministrasjon_application" {
│ 
│ Application.skiperator.kartverket.no "nrl-brukeradministrasjon" is invalid:
│ spec.strategy: Invalid value: "null": spec.strategy in body must be of type
│ object: "null"
╵
╷
│ Error: Kubernetes API Error: Invalid Application.skiperator.kartverket.no [nrl-brukeradministrasjon]
│ 
│   with kubernetes_manifest.nrl-brukeradministrasjon_application,
│   on application.tf line 1, in resource "kubernetes_manifest" "nrl-brukeradministrasjon_application":
│    1: resource "kubernetes_manifest" "nrl-brukeradministrasjon_application" {
│ 
╵
╷
│ Error: spec.strategy
│ 
│   with kubernetes_manifest.nrl-brukeradministrasjon_application,
│   on application.tf line 1, in resource "kubernetes_manifest" "nrl-brukeradministrasjon_application":
│    1: resource "kubernetes_manifest" "nrl-brukeradministrasjon_application" {
│ 
│ Invalid value: "null": spec.strategy in body must be of type object: "null"
```

This field is not a mandatory field.

**Comment by alexsomesan** [comment]
@anderssonw thanks for the additional detail.
TL;DR; some of the operations you are describing are against Kubernetes' rules for managing CRD schemas.

According to Kubernetes' own CRD versioning guidelines, modifying the schema of an existing CRD version while CRs of that version are present in the cluster goes against the rules of CRD safe practices. The errors that you've shared in your most recent reply are in fact coming from the API server itself that is complaining about schema incompatibilities with the stored copy of that CR.
In general, when you make changes to the schema of a CRD, you should be following a set of rules (described in the K8s docs I will link below), but when doing so against a running system, you should have migration web-hooks in place to ensure a predictable conversion between a manifest's schema version and the current API storage version. The safest approach is to increment the API version number of the CRD whenever you make a schema change. If you want to make a schema change to an existing API version, of which you have CR objects present in the cluster, those CRs should be recreated - they will NOT be automatically migrated unless you have explicitly installed migration web-hooks on the API server.

You can find a good summary of the guidelines in this blog post: https://faun.dev/c/stories/dineshparvathaneni/kubernetes-crd-versioning-for-operator-developers/

Full docs on CRD schema versioning are here: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/

We will look into making the provider a bit more robust in cases like this. Right now, it assumes the user is following these practices and doesn't have too many sanity checks in place against that.

**Comment by anderssonw** [comment]
I understand. It's not a huge surprise that it was our fault and not yours. I agree that some sanity checks should be in place. Would've helped us at least!

Thanks for sending the documentation and tips, we will look into this and see if we can't start modifying our CRD the way we should.

However, the last error I sent you was not due to a modified CRD. In fact this product team simply added the `spec.strategy` field, and then removed it. Removing it provided the error I posted. When they added it back with a new value, it worked fine. This has never been a mandatory field in our CRD.

Should I create a separate issue for this, or would you care to look into it here? :)

@alexsomesan
